base/bucket_gocb.go

```diff
@@ -1053,29 +1092,96 @@ func (bucket CouchbaseBucketGoCB) GetWithXattr(k string, xattrKey string, rv int
 
 }
 
-// Delete a document and it's associated named xattr.  Couchbase server will preserve system xattrs as part of the (CBS) tombstone when a document is deleted.
-// To remove the system xattr as well, an explicit subdoc delete operation is required.
+// Only called for Purge
 func (bucket CouchbaseBucketGoCB) DeleteWithXattr(k string, xattrKey string) error {
 
+	return bucket.deleteWithXattrInternal(k, xattrKey, nil)
+
+}
+
+// A function that will be called back after the first delete attempt but before second delete attempt
+type deleteWithXattrRaceInjection func(bucket CouchbaseBucketGoCB, k string, xattrKey string)
+
+// Delete doc + xattrs (purge).  Algorithm:
+// Try to delete body + xattr in single op
+// If fails
+// 		- Do get w/ xattr (get cas)
+// 		- Cas-safe delete just the xattr
+// If that fails with a cas error, return error (maybe race where someone resurrected doc)
+func (bucket CouchbaseBucketGoCB) deleteWithXattrInternal(k string, xattrKey string, callback deleteWithXattrRaceInjection) error {
+
 	bucket.singleOps <- struct{}{}
 	defer func() {
 		<-bucket.singleOps
 	}()
 	gocbExpvars.Add("Delete", 1)
-	removeCas, err := bucket.Bucket.Remove(k, 0)
-	if err != nil && err != gocb.ErrKeyNotFound {
+
+	LogKeys["CRUD+"] = true
+
+	LogTo("CRUD+", "DeleteWithXattr called with key: %v xattrKey: %v", k, xattrKey)
+
+	// This flag seems to work well no matter what the current document state is in
+	deleteFlags := gocb.SubdocDocFlagAccessDeleted
```

Comment:
I've removed `SubdocDocFlagAccessDeleted` everywhere except for the case where we are deleting just the xattr, after it's been determined the doc has no body